File: matlab/dc_main_analysis_obj.m



clear
rng(1)

folder ='/Users/williamviolette/Documents/Philippines/phil_analysis/phil_temp_pay/moments/';
cd_dir ='/Users/williamviolette/Documents/Philippines/phil_analysis/phil_codes_pay/paper/tables/';



real_data      = 1    ;
given_sim      = 1    ;

est_pattern    = 0 ;
results        = 0    ;
boot           = 0    ;
br             = 10   ; % reps

int_size = 1; % number of interpolations
refinement      = 1 ; % refine the number of stuff
one_price       = 0 ;

marginal_cost = 5;
ppinst = 51;
        
s=32*12; % sets account length

mult_set = [  1  ];

n  = 384*50 +1 ; 
rng(1);
X=rand(n-1,2);
sigA = 0;
sigB = 0;
nD   = 2;

ver = 'b';

  %  alpha pd pc 
option = [ 7 12 17 ];   %%% what to estimate
lb = [ 40 10  .01 ];
ub = [ 80 400 .99 ];

option_moments       = [ 1 2 3 4 5  ];  %%% moments to use!
option_moments_est   = [ 1 2 3 4 5  ];

inc_t=1;

    visit_price = 200;
    erate = 45;
    popadj = 2.4*12;
    
[c_avg,c_std,bal_avg,bal_med,bal_std,bal_corr,...
    dc_shr,...
    bal_0,bal_end,bal_0_end,...
    am_d, am_d4,...
    am1,am2,am3,am4,...
    amar1,amar2,amar3,amar4,...
    y_avg,y_cv,Aub,Alb,Blb,...
    p1,p2,prob_caught,...
    delinquency_cost,r_lend,dc_prob] ...
        = import_to_matlab_t3(folder,one_price,1);
    
data_moments = [ c_avg; bal_avg; dc_shr; am_d; bal_0; bal_end ] ;

format long g

    nA = 40
    nB = 40

Alb = -2.*y_avg ;
Aub =  1.*y_avg ;

%%% annual rate of 5.75%, which implies a monthly interest rate of .0047
r_lend     = .0047 ;
r_high     = .0945 ;

% beta_set = .02508  % 1/((1+beta_set)^(12))
beta_set   = .01
if strcmp(ver,'bhigh')==1
    beta_set = .01
end
if strcmp(ver,'blow')==1
    beta_set = .0025
end
% (1+.01)^(12)-1   % 1/(1+x) = 1/((1+d)^12)

% y_cv=.2

    %             1       2        3         4         5       6      7     8         9      10  11   12   13  14     15    16     17     18      19    20  21                
    % given :  r_lend , r_water, r_high, hasscost, inc shock, untie, alpha, beta_up , Y   ,  p1, p2 , pd,  n, curve, fee,  vhass   pc     pm      Blb   Tg  sp     
given =        [   0     0       r_high      0        y_cv      0      50   beta_set  y_avg  p1  p2   413  n    1     0     0    .042  bal_0_end  Blb   12  .8 ];
if strcmp(ver,'bhigh')==1
    given =    [   0     0       r_high      0        y_cv      0      54   beta_set  y_avg  p1  p2   190  n    1     0     0   .24   bal_0_end Blb   ];
end
if strcmp(ver,'blow')==1
    given =    [   0     0       r_high      0        y_cv      0      55   beta_set  y_avg  p1  p2   150  n    1     0     0   .16   bal_0_end Blb   ];
end


% csvwrite(strcat(folder,'given.csv'),given);
                        
            
if real_data == 1
            data = data_moments(option_moments,:); % need to transpose here
else
            data = h(option_moments,:);
end
   
tic
% [est_mom,ucon,controls,~,~,A1,B1]=obj(given,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X);
[est_mom,ucon,controls,~,~,A1,B1]=obj_tgr(given,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X);
toc

disp ' A loan '
sum(controls(:,2)==min(controls(:,2)))
disp ' A savings '
sum(controls(:,2)==max(controls(:,2)))
disp ' B loan '
sum(controls(:,3)==min(controls(:,3)))
disp ' B loan (pre DC)'
sum(controls(controls(:,6)<300,3)==min(controls(controls(:,6)<300,3)))
disp ' B loan (last DC)'
sum(controls(controls(:,6)==s,3)==min(controls(controls(:,6)==s,3)))/sum(controls(:,6)==s)
disp 'B loan 0 (last DC)'
sum(controls(controls(:,6)==s,3)==0)/sum(controls(:,6)==s)
disp ' B loan average '
mean(controls(controls(:,6)==s,3))

disp ' Sim '
round(est_mom(option_moments_est),2)
disp ' Data '
round(data(option_moments),2)
       


if est_pattern==1
        options = optimoptions('patternsearch','Display','iter','MaxFunctionEvaluations',200,'MaxIterations',30,'InitialMeshSize',1,'UseParallel',true);
        weights =  eye(size(data,1))./(data.^2) ;   % normalize moments to be between zero and one (matters quite a bit)
        ag = given(option);    
        obj_run = @(a1)objopt(a1,given,data,option,option_moments_est,weights,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X);
                    disp ' old obj: ' 
                    obj_run(ag)
                    disp ' '
                    disp 'pattern search ... '
                    tic
                    [res,fval,~,Output] = patternsearch(obj_run,ag,[],[],[],[],lb,ub,[],options)
                    fprintf('The number of iterations was : %d\n', Output.iterations);
                    fprintf('The number of function evaluations was : %d\n', Output.funccount);
                    toc
                    [~,~,est_mom]=obj_run(res);
                    disp   '   truth               estimates   ' 
                    [ round(data(:,1),2)  round(est_mom,2) ]
                    disp ' psearch done ! :)'
         csvwrite(strcat(folder,'pattern_estimates_',ver,'.csv'),res)
         
    rb=zeros(br,size(option_moments,2));
    
    if boot==1
        for i=1:size(rb,1)
            rng(i);
            X1=rand(n-1,1);
            c_avg     = csvread(strcat(folder,'c_avg_',num2str(i),'.csv'))  ;
            bal_avg   = csvread(strcat(folder,'bal_avg_',num2str(i),'.csv'));
            dc_shr = csvread(strcat(folder,'dc_shr_',num2str(i),'.csv'));
            data_moments_boot = [ c_avg; bal_avg; dc_shr ]

            data = data_moments_boot(option_moments,:);

            options = optimoptions('patternsearch','Display','iter','MaxFunctionEvaluations',200,'MaxIterations',30,'InitialMeshSize',1,'UseParallel',true);
            weights =  eye(size(data,1))./(data.^2) ;   % normalize moments to be between zero and one (matters quite a bit)
            ag = given(option);    
            obj_run = @(a1)objopt(a1,given,data,option,option_moments_est,weights,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X1);
                        disp ' old obj: ' 
                        obj_run(ag)
                        disp ' '
                        disp 'pattern search ... '
                        tic
                        [res,fval,~,Output] = patternsearch(obj_run,ag,[],[],[],[],lb,ub,[],options)
                        fprintf('The number of iterations was : %d\n', Output.iterations);
                        fprintf('The number of function evaluations was : %d\n', Output.funccount);
                        toc
                        [~,~,est_mom]=obj_run(res);
                        disp   '   truth               estimates   ' 
                        [ round(data(:,1),2)  round(est_mom,2) ]
                        disp ' psearch done ! :)'
             csvwrite(strcat(folder,'pattern_estimates_',ver,'_',num2str(i),'.csv'),res)
        end
    end 
end


    

if results==1
    rb=zeros(br,size(option_moments,2));
    if boot==1
        for i = 1:size(rb,1)
           rb(i,:) =  csvread(strcat(folder,'pattern_estimates_',ver,'_',num2str(i),'.csv'));
        end
    end
    r = csvread(strcat(folder,'pattern_estimates_',ver,'.csv'));

%     j=print_estimates(cd_dir,r,rb,ver);
%     j=fit_print(cd_dir,r,ver,given,option,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X);
if given_sim==1
    r= given(option);
end
    j=counterfactuals_fixedcost(cd_dir,r,ver,given,option,ppinst,r_lend,visit_price,marginal_cost,p1,p2,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X);
end






    
    %%%% BALLPARKING REVENUE RAISING MORE %%%%
    
%    jj = -(-alpha + 2*p2*(alpha^2 - 4*R - 4*R*p2)^(1/2) + (alpha^2 - 4*R - 4*R*p2)^(1/2))/(2*(p2 + 1))
%    jj = -(4*R - p1*(alpha^2 + 2*alpha*p1 + p1^2 - 8*R)^(1/2) - alpha^2 + p1^2 + alpha*(alpha^2 + 2*alpha*p1 + p1^2 - 8*R)^(1/2))/(8*R)
%  
%  price=[1:40]';
%  WSE = (55-price)./(p2.*2.0+1.0);
%  REV = WSE.*(price+p2*WSE);
%  
%  plot(price,REV)
%  
%  REV=zeros(size(price,1),size(price,2))

%  price2=.01.*[1:40]'
%  WSE = (55-p1)./(price2.*2.0+1.0)
%  REV = WSE.*(p1+price2.*WSE)
%  plot(price2,REV)
%  wse = (50-20)./(5*2.0+1.0)
    


%     syms p1s p2s as R
%      assume(p1s>0)
%      assume(p2s>0)
%      assume(as>0)
%      assume(R>0)
%      wse = (as-p1s)./(p2s.*2.0+1.0)
%  rr = wse*(p1s+p2s*wse)
%     solve(rr-R,p1s)
% 
%  assume(p1s>0)
%  assume(p2s>0)
%  assume(as>0)
%  assume(R>0)
%  wse = (as-p1s)./(p2s.*2.0+1.0)
%  rr = wse*(p1s+p2s*wse)
%  solve(rr-R,p2s)



%%% SUPER USEFUL GRID TEST THAT DEFINITELY CONFIRMS 50 50 AND EVEN 40 40
% gt = [20:10:60]' ;
% res = zeros(size(gt,1),5)
% for i=1:size(gt,1)
%     nA = gt(i) 
%     nB = gt(i)
%     tic
%     [est_mom,ucon,controls,~,~,A1,B1]=obj(given,nA,sigA,Alb,Aub,nB,sigB,nD,s,int_size,refinement,X);
%     toc
%     res(i,:)=est_mom';
% end
% res

%         weights =  eye(size(data,1))./(data.^2) ;   % normalize moments to be between zero and one (matters quite a bit)
%         ag = given(option);    
%         obj_run = @(a1)objopt(a1,given,data,option,option_moments_est,weights,nA,sigA,Alb,Aub,nB,sigB,Blb,nD,s,int_size,refinement);
%         obj_run(ag)
                    










% if short_est==1
%         options = optimoptions('surrogateopt','Display','iter','MaxFunctionEvaluations',20,'InitialPoints',given(option));
%         weights =  eye(size(data,1))./(data.^2) ;   % normalize moments to be between zero and one (matters quite a bit)
%         ag = given(option);    
%         obj_run = @(a1)objopt(a1,given,data,option,option_moments_est,weights,nA,sigA,Alb,Aub,nB,sigB,Blb,nD,s,int_size,refinement);
%                     disp ' old obj: ' 
%                     obj_run(ag)
%                     disp ' '
%                     disp 'surrogate opt ... '
%                     tic
%                     [res,fval,~,Output] = surrogateopt(obj_run,lb,ub,options)
%                     toc
%                     [~,~,est_mom]=obj_run(res);
%                     disp   '   truth               estimates   ' 
%                     [ round(data(:,1),2)  round(est_mom,2) ]
%                     disp ' psearch done ! :)'
% end
% 
% 

File: matlab/py/dc_main_analysis_obj.py
#!/usr/bin/env python3
"""
Translate of dc_main_analysis_obj.m — main estimation script.

Usage:
    python -m matlab.py.dc_main_analysis_obj          (from repo root)
    python dc_main_analysis_obj.py                    (from matlab/py/)

Reproduces the MATLAB WBER-version estimation pipeline:
  1. Load data moments from CSV
  2. Run single obj() evaluation and display results
  3. (Optional) Run pattern-search estimation with bootstrap
"""
import sys
import os
import time
import numpy as np

# ---- Repo root detection ----
# Works from: command line, Spyder, Jupyter, interactive REPL.
# Override: set PHIL_CODES_PAY environment variable, or edit _MANUAL_ROOT below.
_MANUAL_ROOT = None   # e.g. '/Users/willviolette/.../phil_codes_pay'

def _find_repo_root():
    # 1. Manual override
    if _MANUAL_ROOT is not None:
        return _MANUAL_ROOT
    # 2. Environment variable
    env = os.environ.get('PHIL_CODES_PAY')
    if env and os.path.isdir(env):
        return env
    # 3. __file__ is available → go up two levels from matlab/py/
    try:
        here = os.path.dirname(os.path.abspath(__file__))
        candidate = os.path.abspath(os.path.join(here, '..', '..'))
        if os.path.isdir(os.path.join(candidate, 'moments')) or \
           os.path.isdir(os.path.join(candidate, 'matlab')):
            return candidate
    except NameError:
        pass
    # 4. Search upward from cwd
    d = os.path.abspath(os.getcwd())
    for _ in range(10):
        if os.path.isdir(os.path.join(d, 'moments')) or \
           os.path.isdir(os.path.join(d, 'matlab', 'py')):
            return d
        d = os.path.dirname(d)
    raise FileNotFoundError(
        "Cannot find repo root. Set _MANUAL_ROOT in this file, or set "
        "the PHIL_CODES_PAY environment variable to the repo path.")

_repo_root = _find_repo_root()
if _repo_root not in sys.path:
    sys.path.insert(0, _repo_root)

from matlab.py.import_data import import_to_matlab_t3
from matlab.py.obj import obj
from matlab.py.objopt import objopt
from matlab.py.grid_int_full import grid_int_full


def main():
    np.random.seed(1)

    # ---- Paths ----
    moments_folder = os.path.join(_repo_root, 'moments')
    tables_dir     = os.path.join(_repo_root, 'paper', 'tables_new')

    # ---- Flags ----
    real_data    = 1
    given_sim    = 1
    est_pattern  = 0
    results      = 1
    boot         = 1
    br           = 10      # bootstrap reps
    opt_method   = 'Nelder-Mead'  # 'Nelder-Mead' or 'Powell'

    int_size    = 1
    refinement  = 1
    one_price   = 0

    marginal_cost = 5
    ppinst        = 51

    s = 32 * 12            # account length = 384

    n  = 384 * 50 + 1      # 19201
    np.random.seed(1)
    X = np.random.rand(n - 1, 2)

    sigA = 0
    sigB = 0
    nD   = 2

    # What to estimate:  alpha(idx 6), pd(idx 11), pc(idx 16)  [0-based]
    option = [6, 11, 16]   # MATLAB [7, 12, 17] → Python 0-based
    lb = np.array([40.0,  10.0, 0.01])
    ub = np.array([80.0, 400.0, 0.99])

    option_moments     = [0, 1, 2]   # MATLAB [1,2,3] → 0-based
    option_moments_est = [0, 1, 2]

    # ---- Load data ----
    d = import_to_matlab_t3(moments_folder, one_price, 1)

    c_avg       = d['c_avg']
    bal_avg     = d['bal_avg']
    dc_shr      = d['dc_shr']
    am_d        = d['am_d']
    bal_0       = d['bal_0']
    bal_end     = d['bal_end']
    bal_0_end   = d['bal_0_end']
    y_avg       = d['y_avg']
    y_cv        = d['y_cv']
    p1          = d['p1']
    p2          = d['p2']
    prob_caught = d['prob_caught']
    Blb         = d['Blb']

    data_moments = np.array([c_avg, bal_avg, dc_shr, am_d, bal_0, bal_end])

    nA = 40
    nB = 40

    Alb = -2.0 * y_avg
    Aub =  2.0 * y_avg

    r_lend = 0.0047
    r_high = 0.0945

    # ---- Loop over sensitivity specifications ----
    for L in [0]:
        ver = 'b'
        if L == 1: ver = 'bhigh'
        if L == 2: ver = 'blow'
        if L == 3: ver = 'chigh'
        if L == 4: ver = 'clow'

        beta_set = 0.005
        if ver == 'bhigh': beta_set = 0.01
        if ver == 'blow':  beta_set = 0.0025

        # MATLAB pattern search estimates (warm-start)
        matlab_est = {'alpha': 54.001, 'pd': 319.56, 'pc': 0.22015}

        #                 0        1        2       3       4       5       6      7         8     9   10    11  12  13   14   15    16       17       18  19   20
        # given:      r_lend, r_water, r_high, hcost, inc_sh, untie, alpha, beta, Y,       p1,  p2,  pd,   n, curve,fee, vh,  pc,   pm,      Blb,  Tg,  sp
        given = np.array([
            0,    0,    r_high, 0,   y_cv,  0,    matlab_est['alpha'],   beta_set, y_avg, p1, p2,  matlab_est['pd'],  n, 1,   0,   0,  matlab_est['pc'], bal_0_end, Blb, 12, 0.8
        ], dtype=np.float64)

        if ver == 'bhigh':
            given = np.array([0,0,r_high,0,y_cv,0,54,  beta_set,y_avg,p1,p2,370,n,1,0,0,0.24,bal_0_end,Blb,12,0.8])
        if ver == 'blow':
            given = np.array([0,0,r_high,0,y_cv,0,54,  beta_set,y_avg,p1,p2,340,n,1,0,0,0.20,bal_0_end,Blb,12,0.8])
        if ver == 'chigh':
            given = np.array([0,0,r_high,0,y_cv,0,53.5,beta_set,y_avg,p1,p2,380,n,2,0,0,0.215,bal_0_end,Blb,12,0.8])
        if ver == 'clow':
            given = np.array([0,0,r_high,0,y_cv,0,54.5,beta_set,y_avg,p1,p2,310,n,0.5,0,0,0.23,bal_0_end,Blb,12,0.8])

        if real_data == 1:
            data = data_moments[option_moments]
        else:
            raise RuntimeError("real_data == 0 not supported")

        # ---- Single obj evaluation ----
        print(f"\n{'='*60}")
        print(f"Running obj() evaluation  [ver={ver}]")
        print(f"{'='*60}")

        t0 = time.time()
        est_mom, ucon, controls, nA_out, nB_out, A1, B1 = \
            obj(given, nA, sigA, Alb, Aub, nB, sigB, nD, s,
                int_size, refinement, X)
        elapsed = time.time() - t0
        print(f"obj() completed in {elapsed:.2f} seconds")

        print(f"\n A loan:         {np.sum(controls[:, 1] == controls[:, 1].min())}")
        print(f" A savings:      {np.sum(controls[:, 1] == controls[:, 1].max())}")
        print(f" B loan:         {np.sum(controls[:, 2] == controls[:, 2].min())}")
        pre_dc = controls[:, 5] < 300
        print(f" B loan (pre DC):{np.sum(controls[pre_dc, 2] == controls[pre_dc, 2].min())}")
        end_dc = controls[:, 5] == s
        if np.any(end_dc):
            print(f" B loan (last DC):{np.sum(controls[end_dc, 2] == controls[end_dc, 2].min()) / np.sum(end_dc):.4f}")
            print(f" B loan 0 (last DC):{np.sum(controls[end_dc, 2] == 0) / np.sum(end_dc):.4f}")
            print(f" B loan average:  {np.mean(controls[end_dc, 2]):.4f}")

        print(f"\n Sim:  {np.round(est_mom[option_moments_est], 3)}")
        print(f" Data: {np.round(data[option_moments], 3)}")

        # ---- Pattern search estimation ----
        if est_pattern == 1:
            from scipy.optimize import minimize

            # Pre-compute grid (doesn't change during optimization)
            precomputed = grid_int_full(
                nA, sigA, Alb, Aub, nB, sigB, Blb, nD,
                int_size, refinement, int(given[5]))

            weights = np.diag(1.0 / data ** 2)
            ag = given[option]

            def obj_fn(a1):
                fval, _, _ = objopt(
                    a1, given, data, option, option_moments_est, weights,
                    nA, sigA, Alb, Aub, nB, sigB, nD, s,
                    int_size, refinement, X, precomputed_grid=precomputed)
                return fval

            print(f"\n old obj: {obj_fn(ag):.6f}")
            print(f" {opt_method} search ...")

            t0 = time.time()
            if opt_method == 'Nelder-Mead':
                result = minimize(obj_fn, ag, method='Nelder-Mead',
                                  options={'maxfev': 400, 'maxiter': 200,
                                           'xatol': 0.5, 'fatol': 1e-5,
                                           'adaptive': True, 'disp': True})
            else:
                result = minimize(obj_fn, ag, method='Powell',
                                  bounds=list(zip(lb, ub)),
                                  options={'maxfev': 400, 'maxiter': 30, 'disp': True})
            elapsed = time.time() - t0
            res = np.clip(result.x, lb, ub)  # enforce bounds for Nelder-Mead
            fval = result.fun

            print(f"Iterations: {result.nit}")
            print(f"Function evaluations: {result.nfev}")
            print(f"Elapsed: {elapsed:.1f}s")

            _, _, est_mom_opt = objopt(
                res, given, data, option, option_moments_est, weights,
                nA, sigA, Alb, Aub, nB, sigB, nD, s,
                int_size, refinement, X, precomputed_grid=precomputed)

            print(f"\n   Data              Estimates")
            for i in range(len(data)):
                print(f"   {data[i]:8.3f}         {est_mom_opt[i]:8.3f}")
            print(" search done! :)")

            np.savetxt(os.path.join(moments_folder,
                       f'pattern_estimates_{ver}.csv'),
                       res[None, :], delimiter=',')

            # ---- Bootstrap ----
            if boot == 1:
                for i_boot in range(1, br + 1):
                    np.random.seed(i_boot)
                    X1 = np.random.rand(n - 1, 2)

                    c_avg_b   = np.loadtxt(os.path.join(moments_folder, f'c_avg_{i_boot}.csv'), delimiter=',')
                    bal_avg_b = np.loadtxt(os.path.join(moments_folder, f'bal_avg_{i_boot}.csv'), delimiter=',')
                    dc_shr_b  = np.loadtxt(os.path.join(moments_folder, f'dc_shr_{i_boot}.csv'), delimiter=',')

                    data_boot = np.array([c_avg_b, bal_avg_b, dc_shr_b])
                    data_b = data_boot[option_moments]
                    weights_b = np.diag(1.0 / data_b ** 2)

                    def obj_fn_b(a1, _d=data_b, _w=weights_b, _X=X1):
                        fval_b, _, _ = objopt(
                            a1, given, _d, option, option_moments_est, _w,
                            nA, sigA, Alb, Aub, nB, sigB, nD, s,
                            int_size, refinement, _X, precomputed_grid=precomputed)
                        return fval_b

                    print(f"\n Bootstrap rep {i_boot}")
                    t0 = time.time()
                    if opt_method == 'Nelder-Mead':
                        result_b = minimize(obj_fn_b, ag, method='Nelder-Mead',
                                            options={'maxfev': 400, 'maxiter': 200,
                                                     'xatol': 0.5, 'fatol': 1e-5,
                                                     'adaptive': True, 'disp': True})
                    else:
                        result_b = minimize(obj_fn_b, ag, method='Powell',
                                            bounds=list(zip(lb, ub)),
                                            options={'maxfev': 400, 'maxiter': 30, 'disp': True})
                    print(f"  Elapsed: {time.time()-t0:.1f}s")

                    np.savetxt(os.path.join(moments_folder,
                               f'pattern_estimates_{ver}_{i_boot}.csv'),
                               result_b.x[None, :], delimiter=',')

        # ---- Results summary ----
        if results == 1 and est_pattern == 1:
            rb = np.zeros((br, len(option_moments)))
            if boot == 1:
                for i in range(1, br + 1):
                    rb[i-1, :] = np.loadtxt(
                        os.path.join(moments_folder,
                                     f'pattern_estimates_{ver}_{i}.csv'),
                        delimiter=',')
            r = np.loadtxt(os.path.join(moments_folder,
                           f'pattern_estimates_{ver}.csv'), delimiter=',')
            print(f"\n Estimates: {r}")
            print(f" Bootstrap SE: {rb.std(axis=0)}")

    print("\nDone.")


if __name__ == '__main__':
    main()
